### PR TITLE
Feat: export current env to html

### PR DESCRIPTION
**Pull request by rajnisht7**
## Description
This PR adds a new feature which allow user to download the complete env in html form, that is export it as html and it can be shared, viewed anytime, downloaded and all the operations can be performed on the panes in the html file which can be performed while live server.

## Motivation and Context
sometimes people need to share the visdom dashboard for research papers, team presentations, reports, or when the visdom server is not running, this PR resolves this.

how it works

https://github.com/user-attachments/assets/ed84d1ef-0513-4cd2-b42e-033f5c659f16

sample html file(after downloading)
[visdom_exported.html](https://github.com/user-attachments/files/26710418/visdom_exported.html)

## How Has This Been Tested?
started the server
clicked on download button
verified that the download html file contains all the panes and operations are working properly
also tested with empty env, raises error

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Add the ability to export the current Visdom environment as a standalone HTML dashboard that can be downloaded from the UI.

New Features:
- Introduce an HTML export function that serializes the current environment panes (plots, images, and text) into a self-contained HTML file.
- Add a topbar control button to trigger downloading the current environment as an HTML file when connected and an environment is selected.

### CODE REVIEW

File: js/main.js [hunk 1]
```diff
@@ -769,6 +769,97 @@ const App = () => {
...
+<body>
+  <div class="container">
+    <h1 style="text-align:center; margin-bottom:40px; color:#337ab7;">${title}</h1>
+    <div class="row">`;
+
+    Object.keys(storeData.panes).forEach((id) => {
+      const pane = storeData.panes[id];
+      const layoutItem = storeData.layout.find((l) => l.i === id);
+      if (!layoutItem) return;
+
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: tested this approach, this approach introducing uneven gaps, column and row breaks in the exported html file, therefore keeping the current approach

File: js/main.js [hunk 2]
```diff
@@ -769,6 +769,97 @@ const App = () => {
...
+    const blob = new Blob([html], { type: 'text/html' });
+    const url = URL.createObjectURL(blob);
+    const a = document.createElement('a');
+    a.href = url;
+    a.download = `visdom_${selection.envIDs.join('_')}_${new Date()
+      .toISOString()
+      .slice(0, 19)}.html`;
+    a.click();
+    URL.revokeObjectURL(url);
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: corrected

File: js/main.js [hunk 3]
```diff
@@ -769,6 +769,97 @@ const App = () => {
     }
   });
 
+  const exportCurrentEnvToHtml = () => {
+    if (!storeData.panes || Object.keys(storeData.panes).length === 0) {
+      alert('No pane available to download');
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: will add

File: js/main.js [hunk 4]
```diff
@@ -769,6 +769,97 @@ const App = () => {
...
+          document.getElementById('plot-${id}').innerHTML = '<img src="${pane.content.src}" style="max-width:100%; height:auto; display:block; margin:auto;">';
+        </script>`;
+      }
+      // Text pane
+      else if (pane.type === 'text' && pane.content) {
+        html += `
+        <script>
+          document.getElementById('plot-${id}').innerHTML = \`${pane.content}\`;
+        </script>`;
+      }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: will add

File: js/main.js [hunk 5]
```diff
@@ -769,6 +769,97 @@ const App = () => {
...
+</body>
+</html>`;
+
+    const blob = new Blob([html], { type: 'text/html' });
+    const url = URL.createObjectURL(blob);
+    const a = document.createElement('a');
+    a.href = url;
+    a.download = `visdom_${selection.envIDs.join('_')}_${new Date()
+      .toISOString()
+      .slice(0, 19)}.html`;
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: corrected

File: js/main.js [hunk 6]
```diff
@@ -768,6 +768,130 @@ const App = () => {
...
+            <div class="pane-title">${
+              pane.title || pane.type.toUpperCase()
+            }</div>
+            <div class="content" id="plot-${id}"></div>
+          </div>
+        </div>`;
+
+      // Plot pane
+      if (pane.type === 'plot' && pane.content?.data) {
+        html += `
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: will add

File: js/main.js [hunk 7]
```diff
@@ -768,6 +768,129 @@ const App = () => {
       );
     }
   });
+  const exportCurrentEnvToHtml = () => {
+    if (!storeData.panes || Object.keys(storeData.panes).length === 0) {
+      alert('No Pane available to export');
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: will add

File: js/template/exportTemplate.js [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+  // SVG
+  if (t === 'svg') {
+    const raw = c ? (c.content || (typeof c === 'string' ? c : '')) : '';
+    const d = mkEl('div', 'pad');
+    d.style.textAlign = 'center';
+    d.innerHTML = raw;
+    const s = d.querySelector('svg');
+    if (s) { s.style.maxWidth = '100%'; s.style.height = 'auto'; }
+    pc.appendChild(d);
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:134]: as users shares their own data, not just random, current implementation can work

File: js/template/exportTemplate.js [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+.pad{padding:8px}
+.pc table{width:100%;border-collapse:collapse;font-size:12px}
+.pc td,.pc th{border:1px solid #e5e7eb;padding:4px 8px}
+.pc th{background:#f5f7fa;font-weight:600}
+.note{padding:10px;font-size:11px;color:#999;font-style:italic}
+</style>
+</head>
+<body>
+<h1>${title}</h1>
+<div id="board"></div>
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:49]: corrected

File: js/template/exportTemplate.js [hunk 10]
```diff
@@ -0,0 +1,297 @@
+function buildExportHtml(title, paneData, validIds) {
+  const S = '<' + '/script>';
+  return `<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="utf-8">
+<title>${title}</title>
+<script src="https://cdn.plot.ly/plotly-2.27.0.min.js">${S}
+<style>
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:26]: @marcoag, wanted to ask regarding this comment
 can make it complete offline by inlining the complete plotly library but this doesnt seem to be nice idea, as per me CDN would be better

File: js/template/exportTemplate.js [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+  if (cls) e.className = cls;
+  return e;
+}
+
+function renderContent(id, pane, pc) {
+  const c = pane.content;
+  const t = pane.type;
+
+  // Any Plotly-compatible pane (plot, surface, or unknown with .data)
+  if (c && c.data) {
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js]: made more specific

File: js/template/exportTemplate.js [hunk 12]
```diff
@@ -0,0 +1,297 @@
+function buildExportHtml(title, paneData, validIds) {
+  const S = '<' + '/script>';
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:19]: added

File: js/main.js [hunk 13]
```diff
@@ -768,6 +769,57 @@ const App = () => {
...
+    const blob = new Blob([html], { type: 'text/html' });
+    const url = URL.createObjectURL(blob);
+    const a = document.createElement('a');
+    a.href = url;
+    a.download = `visdom_${selection.envIDs.join('_')}_${safeTs}.html`;
+    document.body.appendChild(a);
+    a.click();
+    document.body.removeChild(a);
+    setTimeout(() => URL.revokeObjectURL(url), 100);
+  };
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:886]: will add

File: js/topbar/ViewControls.js [hunk 14]
```diff
@@ -81,6 +82,16 @@ function ViewControls(props) {
...
+        <button
+          data-toggle="tooltip"
+          title="Export as HTML"
+          data-placement="bottom"
+          className="btn btn-default"
+          disabled={!(connected && envIDs.length > 0)}
+          onClick={onExportHtml}
+        >
+          <span className="glyphicon glyphicon-download" />
+        </button>
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/topbar/ViewControls.js:95]: added

File: js/template/exportTemplate.js [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+  }
+
+  // SVG
+  if (t === 'svg') {
+    const raw = c ? (c.content || (typeof c === 'string' ? c : '')) : '';
+    const d = mkEl('div', 'pad');
+    d.style.textAlign = 'center';
+    d.innerHTML = raw;
+    const s = d.querySelector('svg');
+    if (s) { s.style.maxWidth = '100%'; s.style.height = 'auto'; }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:133]: as people create their own panes, therefore chance of sending embeddings intentionally is quite low therefore keeping current implementation and the suggestion can also make some advance svg features limited

File: js/template/exportTemplate.js [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+    d.appendChild(img); pc.appendChild(d);
+    return;
+  }
+
+  // Text
+  if (t === 'text') {
+    const d = mkEl('div', 'pad');
+    d.style.fontSize = '13px';
+    d.innerHTML = c ? String(c) : '';
+    pc.appendChild(d);
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:122]: sanitizing can break the expected output, hence keeping the current implementation

File: js/template/exportTemplate.js [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+function buildExportHtml(title, paneData, validIds) {
+  const safeData = JSON.stringify(paneData).replace(/<\/script/gi, '<\\/script');
+  const safeIds  = JSON.stringify(validIds).replace(/<\/script/gi, '<\\/script');
+  const S = '<' + '/script>';
+  return `<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="utf-8">
+<title>${title}</title>
+<script src="https://cdn.plot.ly/plotly-2.27.0.min.js">${S}
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:25]: already done in `main.js`

File: js/template/exportTemplate.js [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+    });
+
+    const ex = [], ey = [];
+    edges.forEach((e) => {
+      const f = pos[e.from], t2 = pos[e.to];
+      if (f && t2) {
+        ex.push(f.x, t2.x, null);
+        ey.push(f.y, t2.y, null);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:212]: tested the current implementation,no issue have been observed at the moment, have also updated the implementation to support both formats

File: js/main.js [hunk 19]
```diff
@@ -768,6 +769,57 @@ const App = () => {
       );
     }
   });
+  const exportCurrentEnvToHtml = () => {
+    if (!storeData.panes || Object.keys(storeData.panes).length === 0) {
+      alert('No Pane Available to export');
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: will add

File: js/template/exportTemplate.js [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+    }
+ 
+    else if (c && Array.isArray(c.X) && c.X.length) {
+      xs = c.X.map((p) => p[0]);
+      ys = c.X.map((p) => p[1]);
+      if (!labels.length && c.Y) labels = c.Y;
+    }
+
+    if (xs && xs.length) {
+      const w = mkEl('div');
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:268]: tested with current data and embeddings export works properly
can add support for `content.data[].position` if this format expected to be used

File: js/main.js [hunk 21]
```diff
@@ -768,6 +769,57 @@ const App = () => {
...
+
+    const validIds = sortedIds.filter((id) => paneData[id]);
+
+    const html = buildExportHtml(title, paneData, validIds);
+
+    const blob = new Blob([html], { type: 'text/html' });
+    const url = URL.createObjectURL(blob);
+    const a = document.createElement('a');
+    a.href = url;
+    a.download = `visdom_${selection.envIDs.join('_')}_${safeTs}.html`;
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js]: corrected

File: cypress/integration/export.js [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+    cy.window().then((win) => {
+      cy.stub(win, 'alert').as('alertStub');
+    });
+
+    cy.get(exportButton).should('not.be.disabled').click();
+
+    cy.get('@alertStub').should(
+      'have.been.calledWith',
+      'No panes available to export.'
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply cypress/integration/export.js:90]: corrected

File: js/main.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: added

File: js/main.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: added

File: js/main.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: added

File: js/main.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:886]: added

File: js/main.js [hunk 23]
```diff
@@ -768,6 +769,72 @@ const App = () => {
...
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&#39;');
+  };
+  const exportCurrentEnvToHtml = () => {
+    if (!storeData.panes || Object.keys(storeData.panes).length === 0) {
+      alert('No panes available to export.');
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:833]: added the test and verified as wel

File: js/main.js [hunk 24]
```diff
@@ -813,6 +880,7 @@ const App = () => {
       }}
       onViewChange={updateToLayout}
       onViewManageButton={() => setShowViewModal(!showViewModal)}
+      onExportHtml={exportCurrentEnvToHtml}
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:943]: added

File: js/template/exportTemplate.js [hunk 25]
```diff
@@ -0,0 +1,322 @@
...
+    /<\/script/gi,
+    '<\\/script'
+  );
+  const S = '<' + '/script>';
+  return `<!DOCTYPE html>
+<html lang="en">
+<head>
+<meta charset="utf-8">
+<title>${title}</title>
+<script src="https://cdn.plot.ly/plotly-2.27.0.min.js">${S}
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:25]: already resolved in `main.js`

File: js/template/exportTemplate.js [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+    return;
+  }
+
+  // Image history
+  if (t === 'image_history') {
+    let src = c && c.src;
+    if (!src && Array.isArray(c) && c.length) {
+      const idx = typeof pane.selected === 'number'
+        ? pane.selected : c.length - 1;
+      const item = c[Math.max(0, idx)];
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js]: corrected

**Comment by Manik-Khajuria-5** [comment]
@rajnisht7 Please address co pilot reviews only after that I can review ur PR

**Review by Jayantparashar10 — APPROVED** [review]
Tested locally — ran `plot_line_basic` and `text_basic` demos, and clicked the export button. The HTML file downloads correctly and renders all panes with working Plotly interactivity.

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@rajnisht7  In exportTemplate.js line 139 and 148 the text and svg content is directly put into innerHTML without cleaning it. so if someone sends something like

  vis.text(< onerror="steal cookies">) to the server and later someone exports that env as html and opens it, the malicious code just runs.

**Comment by rajnisht7** [comment]
> @rajnisht7 In exportTemplate.js line 139 and 148 the text and svg content is directly put into innerHTML without cleaning it. so if someone sends something like
> 
> vis.text(< onerror="steal cookies">) to the server and later someone exports that env as html and opens it, the malicious code just runs.

tested exporting this
```
from visdom import Visdom

vis = Visdom(env='1198')

vis.text(
    '< onerror="steal cookies">'
)
```
[visdom_1198_2026-06-02_17-24.html](https://github.com/user-attachments/files/28519878/visdom_1198_2026-06-02_17-24.html)


have also attached the html file 
works fine

**Review by tonypzy — COMMENTED** [review]
When I execute `exmaple/demo.py` and clicked the button "EXPORT AS HTML", the size of downloaded is approximately 35.8 MB. I think it is unaffordable and too big. Do you have any ideas to reduce the size of downloaded html?

Additionally, could we slightly increase the icon size on the export button and ensure it's properly centered? Right now, it feels a bit out of proportion compared to the adjacent 'Repack' and 'Manage Views' buttons.

**Comment by tonypzy** [comment]
By the way, could you run `npm run build` to update `main.js` and `main.js.map`, and push the changes? The JS files were modified in this branch.

**Comment by rajnisht7** [comment]
> When I execute `exmaple/demo.py` and clicked the button "EXPORT AS HTML", the size of downloaded is approximately 35.8 MB. I think it is unaffordable and too big. Do you have any ideas to reduce the size of downloaded html?
> 
> Additionally, could we slightly increase the icon size on the export button and ensure it's properly centered? Right now, it feels a bit out of proportion compared to the adjacent 'Repack' and 'Manage Views' buttons.

@tonypzy already had a discussion about the same and decided to continue without compressing to keep the plot quality

**Review by vedansh-5 — APPROVED** [review]
LGTM!

File: js/template/exportTemplate.js [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
+    const img = mkEl('img'); img.src = String(src);
+    d.appendChild(img); pc.appendChild(d);
+    return;
+  }
+
+  // Text
+  if (t === 'text') {
+    const d = mkEl('div', 'pad');
+    d.style.fontSize = '13px';
+    d.innerHTML = c ? String(c) : '';
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/template/exportTemplate.js:121]: for dev tool only possible if developer do it deliberately

File: js/main.js [hunk 28]
```diff
@@ -821,6 +822,72 @@ const App = () => {
...
+        content: pane.content,
+        selected: pane.selected,
+        initW: Math.max(280, Math.round(w2p(li.w))),
+        initH: Math.max(200, Math.round(h2p(li.h))),
+      };
+    });
+
+    const validIds = sortedIds.filter((id) => paneData[id]);
+
+    const html = buildExportHtml(title, paneData, validIds);
```
[the remark this answers is not in the document]
Comment (rajnisht7) [reply js/main.js:873]: already have gurad if have zero panes

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@rajnisht7 Can u please test for vis.text(img onerror="steal_cookies()" src=x) I was referring this above

**Comment by rajnisht7** [comment]
> vis.text(img onerror="steal_cookies()" src=x)

@Manik-Khajuria-5 exported properly 
[visdom_1198_2026-06-03_12-24.html](https://github.com/user-attachments/files/28550674/visdom_1198_2026-06-03_12-24.html)

**Review by Manik-Khajuria-5 — APPROVED** [review]
@rajnisht7 Nice work! LGTM

File: js/template/exportTemplate.js [hunk 29]
```diff
@@ -0,0 +1,323 @@
...
+  if (t === 'svg') {
+    const raw = c ? (c.content || (typeof c === 'string' ? c : '')) : '';
+    const d = mkEl('div', 'pad');
+    d.style.textAlign = 'center';
+    d.innerHTML = raw;
+    const s = d.querySelector('svg');
+    if (s) { s.style.maxWidth = '100%'; s.style.height = 'auto'; }
+    pc.appendChild(d);
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (Saksham-Sirohi) [reply js/template/exportTemplate.js:136]: @rajnisht7 is this case handled

**Review by Saksham-Sirohi — APPROVED** [review]
